File: joi2008ho/joi2008ho_c.py
"""
Dart
https://atcoder.jp/contests/joi2008ho/tasks/joi2008ho_c
"""
import bisect

def binary_search(arr, key):
    def is_ok(index):
        return arr[index] <= key

    ok = -1
    ng = len(arr)

    while abs(ok - ng) > 1:
        mid = (ok + ng) // 2
        if is_ok(mid):
            ok = mid
        else:
            ng = mid
    return ok


def solve(p):
    xs = [
        p[i] + p[j]
        for i in range(N + 1)
        for j in range(i, N + 1)
        if p[i] + p[j] <= M
    ]
    xs.sort()
    res = -100
    for x in xs:
        idx = binary_search(xs, M - x)
        # idx = bisect.bisect_right(xs, M - x) - 1
        total = xs[idx] + x
        if res < total:
            res = total

    print("xs:", xs)
    return res


N, M = map(int, input().split())
ps = [int(input()) for _ in range(N)]
ps.append(0)

res = solve(ps)
print(res)

File: joi2008ho/joi2008ho_c.cc
#include <iostream>
#include <cmath>
#include <vector>
#include <map>
#include <set>
#include <string>
#include <cstring>
#include <queue>
#include <ctime>
#include <cassert>
#include <cstdio>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <bitset>

using namespace std;

int binary_search(vector<int> & arr, int key) {

    auto is_ok = [&](int index) {
        return (arr[index] <= key);
    };

    int ok = -1;
    int ng = arr.size();

    while (abs(ok - ng) > 1) {
        int mid = (ok + ng) / 2;
        if (is_ok(mid)) ok = mid;
        else ng = mid;
    }
    return ok;
}

int solve(vector<int> & p, int n, int m) {
    p.push_back(0);

    vector<int> q;
    for (int i = 0; i <= n; ++i) {
        for (int k = i; k <= n; ++k ) {
            int twosum = p[i] + p[k];
            if (twosum <= m) q.push_back(twosum);
        }
    }
    sort(q.begin(), q.end());

    int res = -1000;
    for (auto x: q) {
        int idx = binary_search(q, m - x);
        int total = q[idx] + x;
        if (total > res) res = total;
    }
    return res;
}


int main() {
    int n, m;
    cin >> n >> m;
    vector<int> p(n);
    for (int i = 0; i < n; ++i) {
        cin >> p[i];
    }
    cout << solve(p, n, m) << endl;
}